compiler/rustc_mir_build/src/thir/pattern/usefulness.rs

```diff
@@ -600,191 +619,306 @@ impl<'p, 'tcx> FromIterator<PatStack<'p, 'tcx>> for Matrix<'p, 'tcx> {
     }
 }
 
-/// Represents a set of `Span`s closed under the containment relation. That is, if a `Span` is
-/// contained in the set then all `Span`s contained in it are also implicitly contained in the set.
-/// In particular this means that when intersecting two sets, taking the intersection of some span
-/// and one of its subspans returns the subspan, whereas a simple `HashSet` would have returned an
-/// empty intersection.
-/// It is assumed that two spans don't overlap without one being contained in the other; in other
-/// words, that the inclusion structure forms a tree and not a DAG.
-/// Intersection is not very efficient. It compares everything pairwise. If needed it could be made
-/// faster by sorting the `Span`s and merging cleverly.
-#[derive(Debug, Clone, Default)]
-pub(crate) struct SpanSet {
-    /// The minimal set of `Span`s required to represent the whole set. If A and B are `Span`s in
-    /// the `SpanSet`, and A is a descendant of B, then only B will be in `root_spans`.
-    /// Invariant: the spans are disjoint.
-    root_spans: Vec<Span>,
+/// Given a pattern or a pattern-stack, this struct captures a set of its subpattern branches. We
+/// use that to track unreachable sub-patterns arising from or-patterns. In the absence of
+/// or-patterns this will always be either `Empty` or `Full`.
+/// We support a limited set of operations, so not all possible sets of subpatterns can be
+/// represented. That's ok, we only want the ones that make sense to capture unreachable
+/// subpatterns.
+/// What we're trying to do is illustrated by this:
+/// ```
+/// match (true, true) {
+///     (true, true) => {}
+///     (true | false, true | false) => {}
+/// }
+/// ```
+/// When we try the alternatives of the first or-pattern, the last `true` is unreachable in the
+/// first alternative but no the other. So we don't want to report it as unreachable. Therefore we
+/// intersect sets of unreachable patterns coming from different alternatives in order to figure
+/// out which subpatterns are overall unreachable.
```

Comment:
This paragraph is slightly unclear. "the alternatives of the first or-pattern" indicates to me the two `true | false` patterns, but I'm not sure what "the last `true` is unreachable in the first alternative" means: there's only one `true` in the first alternative `true | false`.
"but no the other" should be "but not the other"